### PR TITLE
Feature getOpenWindows to have list of open windows

### PR DESCRIPTION
**Pull request by miniben-90**
# Why

In my case I need to have access to the active window and the list of open windows.

# Changes

* Some codes moved into functions to avoid duplication of it.
* Add to all systems some codes to have access to the informations.
* getOpenWindows (promise version) and getOpenWindowsSync (sync version) return a `Map<string, {processes: activeWindow.Result[]}>`

### CODE REVIEW

File: Sources/ActiveWinCLI/main.swift [hunk 1]
```diff
@@ -93,12 +82,47 @@ for window in windows {
...
+else {
+	exitWithoutResult()
+}
+
+var openWindows: Array<Any> = [];
+
+for window in windows {
+	let windowOwnerPID = window[kCGWindowOwnerPID as String] as! pid_t // Documented to always exist.
+	let windowInformation = getWindowInformation(window, windowOwnerPID, frontmostAppPID, enableOpenWindowsList)
+	if windowInformation != nil {
```
Comment (sindresorhus) [inline Sources/ActiveWinCLI/main.swift:115]: Use `guard` to reduce nesting.

File: index.d.ts [hunk 2]
```diff
@@ -88,6 +88,8 @@ declare namespace activeWindow {
 	}
 
 	type Result = MacOSResult | LinuxResult | WindowsResult;
+
+	type OpenWindows = Map<string, {processes: Result[]}>;
```
Comment (sindresorhus) [inline index.d.ts:92]:
```suggestion
	type OpenWindows = Result[];
```

Comment (sindresorhus) [reply index.d.ts:92]: Actually. Don't even need the type, just use `Result[]` inline.

File: index.d.ts [hunk 3]
```diff
@@ -144,6 +146,60 @@ declare const activeWindow: {
...
+			console.log(`Process title is ${result.title} with bundle id ${result.owner.bundleId}.`);
+		} else if (result.platform === 'windows') {
+			console.log(`Process title is ${result.title} with path ${result.owner.path}.`);
+		} else {
+			console.log(`Process title is ${result.title} with path ${result.owner.path}.`);
+		}
+	})();
+	```
+	*/
+	getOpenWindows(options?: undefined): Promise<activeWindow.OpenWindows | undefined>;
```
Comment (sindresorhus) [inline index.d.ts:177]: It (the JS code too) should return an empty array, not `undefined`.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
Thanks for working on this. Make sure you look over your own diff and try to improve it. This will save me time so I don't have to point out lots of nitpick stuff.

File: index.d.ts (hunk 2 above)
Comment (sindresorhus) [reply index.d.ts]: Map keys are expected to be unique, but process names are not guaranteed to be unique, so this is not going to work.

Comment (miniben-90) [reply index.d.ts]: I have make the change for all system, now it will return an array only.

**Comment by sindresorhus** [comment]
Maybe you can find some inspiration here: https://github.com/glsorre/amethystwindows/blob/340d54a076b8c04e74b5f096af7c34b5c2f41ba5/AmethystWindows/DesktopWindowsManager/DesktopWindow.cs#L28-L36

**Comment by sindresorhus** [comment]
> Can you be more specific about your links?

The projects I linked to are mature and have a lot of code to handle edge-cases. We can borrow some of that, which are the lines I linked to.

**Comment by miniben-90** [comment]
I make some big breaking change for windows cause of a issue with ffi-napi end Workers. I change it for a native node addon (with node-addon-api, node-gyp).

I also change your main.yml to update version and add node 18 to the test.

I created a release.yml file to generate binary for windows at release (it's important to change version of package.json before a release).

For your previous links, it's not logical to filter with className (example with terminal), I just keep the idea of the last link.

File: package.json [hunk 4]
```diff
@@ -3,7 +3,10 @@
 	"version": "7.7.2",
 	"description": "Get metadata about the active window (title, id, bounds, owner, URL, etc). Works on macOS, Linux, Windows.",
 	"license": "MIT",
-	"repository": "sindresorhus/active-win",
+	"repository": {
+		"type": "git",
+		"url": "git+https://github.com/sindresorhus/active-win.git"
+	},
```
Comment (sindresorhus) [inline package.json:9]: Don't make unrelated changes.

File: readme.md [hunk 5]
```diff
@@ -106,3 +122,12 @@ If you use this package in an Electron app that is sandboxed and you want to get
 
 - [Sindre Sorhus](https://github.com/sindresorhus)
 - [Sebastián Ramírez](https://github.com/tiangolo)
+
+## For development
+
+In case of development to bypass gyp build:
+
+```sh
+yarn install --ignore-scripts
```
Comment (sindresorhus) [inline readme.md:131]: Show the npm command, not yarn.

File: lib/windows-binding.js [hunk 6]
```diff
@@ -0,0 +1,14 @@
+const binary = require('@mapbox/node-pre-gyp');
+const path = require('path');
+const fs = require('fs');
+const bindingPath = binary.find(path.resolve(path.join(__dirname, '../package.json')));
+const binding = (fs.existsSync(bindingPath)) ? require(bindingPath) : {
+	getActiveWindow: () => {
+		return null;
+	},
```
Comment (sindresorhus) [inline lib/windows-binding.js:8]:
```suggestion
	getActiveWindow: () => {},
```

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        with:
+          msbuild-architecture: ${{ matrix.target }}
+
+      - name: Setup node (node=${{ matrix.node }}, host=${{ matrix.host }})
+        uses: actions/setup-node@v3
+        with:
+          node-version: ${{ matrix.node }}
+          architecture: ${{ matrix.host }}
+      - name: Install dependencies
+        run: yarn install --ignore-scripts
```
Comment (sindresorhus) [inline .github/workflows/release.yml:50]: I prefer to use `npm`.

File: .gitignore [hunk 8]
```diff
@@ -3,3 +3,10 @@ yarn.lock
 /.build
 /*.xcodeproj
 /main
+.vscode/settings.json
+.vscode/ipch
+.vscode/c_cpp_properties.json
```
Comment (sindresorhus) [inline .gitignore:8]: These are not project specific and should be in your own global gitignore.

https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

File: Sources/windows/main.cc [hunk 9]
```diff
@@ -0,0 +1,301 @@
+#include <cmath>
+#include <cstdint>
+#include <napi.h>
+#include <iostream>
+#include <shtypes.h>
+#include <string>
+#include <windows.h>
+// #include <filesystem>
```
Comment (sindresorhus) [inline Sources/windows/main.cc:8]: ?

File: Sources/windows/main.cc [hunk 10]
```diff
@@ -0,0 +1,301 @@
...
+  return std::wstring(str.begin(), str.end());
+}
+
+std::string getFileName(const std::string &s)
+{
+  char sep = '/';
+#ifdef _WIN32
+  sep = '\\';
+#endif
+  size_t i = s.rfind(sep, s.length());
```
Comment (sindresorhus) [inline Sources/windows/main.cc:40]: Don't use abbreviated variable names. Applies everywhere.
```suggestion
  size_t index = string.rfind(separator, string.length());
```

File: Sources/windows/main.cc [hunk 11]
```diff
@@ -0,0 +1,301 @@
...
+  int len = WideCharToMultiByte(CP_UTF8, 0, str.c_str(), str.length(), NULL, 0, NULL, NULL);
+  if (len > 0)
+  {
+    ret.resize(len);
+    WideCharToMultiByte(CP_UTF8, 0, str.c_str(), str.length(), &ret[0], len, NULL, NULL);
+  }
+  return ret;
+}
+
+// REturn window title in utf8 string
```
Comment (sindresorhus) [inline Sources/windows/main.cc:61]: Typo

**Comment by sindresorhus** [comment]
This is starting to look good. Although, I'm not super excited to have to maintain C++ code. I would much have preferred Rust and https://napi.rs. Which would have made the code easier to maintain, read, and build.

File: Sources/ActiveWinCLI/main.swift [hunk 12]
```diff
@@ -16,57 +16,43 @@ func exitWithoutResult() -> Never {
...
-	let windowOwnerPID = window[kCGWindowOwnerPID as String] as! pid_t // Documented to always exist.
-
-	if windowOwnerPID != frontmostAppPID {
-		continue
 	}
+	print(string)
+	exit(0)
+}
 
+func getWindowInformation(_ window: [String: Any],_ windowOwnerPID: pid_t,_ frontmostAppPID: pid_t) -> [String: Any]? {
```
Comment (sindresorhus) [inline Sources/ActiveWinCLI/main.swift:27]:
```suggestion
func getWindowInformation(window: [String: Any], windowOwnerPID: pid_t) -> [String: Any]? {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: package.json (hunk 4 above)
Comment (miniben-90) [reply package.json:9]: It's related for the publishing of the binary at release. It's a requirement for the github action.

**Comment by miniben-90** [comment]
> This is starting to look good. Although, I'm not super excited to have to maintain C++ code. I would much have preferred Rust and https://napi.rs. Which would have made the code easier to maintain, read, and build.

Unfortunately, I don't have time to learn a new language at the moment. But you are welcome to make the change on the basis of the main.cc :).

File: .github/workflows/release.yml (hunk 7 above)
Comment (miniben-90) [reply .github/workflows/release.yml:50]: It's more faster then npm. I base the release script on the [node-sqlite3 project](https://github.com/TryGhost/node-sqlite3/blob/master/.github/workflows/ci.yml). This part has no impact on the viability of the project.

Comment (sindresorhus) [reply .github/workflows/release.yml:50]: > It's more faster then npm.

The CI step will be slow anyway. I'm the one stuck maintaining all this code. I don't want to depend on Yarn.

**Comment by sindresorhus** [comment]
@miniben-90 The binary publish step failed: https://github.com/sindresorhus/active-win/actions/runs/3686184879

**Comment by sindresorhus** [comment]
@miniben-90 Different failure: https://github.com/sindresorhus/active-win/actions/runs/3686737172
